src/coreclr/jit/codegenxarch.cpp

```diff
@@ -4456,6 +4444,20 @@ void CodeGen::genCodeForShift(GenTree* tree)
                 GetEmitter()->emitIns_R_ARX(INS_lea, size, tree->GetRegNum(), operandReg, operandReg, 1, 0);
             }
         }
+        // Optimize "X<<2" to "lea [reg*4]" - we only do this when the dst and src registers are different since it will
+        // remove a 'mov'.
+        else if (tree->OperIs(GT_LSH) && !tree->gtOverflowEx() && !tree->gtSetFlags() && shiftBy->IsIntegralConst(2) &&
```

Comment:
Its a bit unfortunate we have all these duplicated checks across these two paths.

Would be nice if we could get `tree->OperIs(GT_LSH) && !tree->gtOverflowEx() && !tree->gtSetFlags()` checked once, same with `tree->GetRegNum() != operandReg`

Reply:
Luckily it's not that much duplication and it won't need to be expanded any further than it already is.

Reply:
Is it worth caching it in a local at least, since refactoring the if statements would be more complicated due to the trailing `else`?

Reply:
I can probably do `tree->OperIs(GT_LSH) && !tree->gtOverflowEx() && !tree->gtSetFlags()` in a single local.